pkg/controllers/externalsecret/externalsecret_controller.go

```diff
@@ -317,7 +317,28 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 		log.V(1).Info("secret creation skipped due to creationPolicy=None")
 		err = nil
 	default:
-		_, err = ctrl.CreateOrUpdate(ctx, r.Client, secret, mutationFunc)
+		op, opErr := ctrl.CreateOrUpdate(ctx, r.Client, secret, mutationFunc)
+		err = opErr
+
+		if op == controllerutil.OperationResultCreated {
+			children := &v1.SecretList{}
+
+			err = r.Client.List(ctx, children,
+				client.InNamespace(externalSecret.Namespace),
+				client.MatchingLabels(map[string]string{
+					esv1beta1.AnnotationSecretOwner: externalSecret.Name,
+				}),
+			)
+
+			for _, secItem := range children.Items {
+				if secItem.Name != secret.Name {
+					err = r.Client.Delete(ctx, &secItem)
+				}
+				if err != nil {
+					break
+				}
+			}
+		}
```

Comment:
wouldn't a list in every update be expensive if we have e.g. 3000 ExternalSecrets?

Reply:
I think what would be better would be (if possible) to filter the list by using the `FieldSelector`, so we can filter out only secrets that are owner-referenced by this ExternalSecrets object. This way we are sure that we can delete any non-matching secret there.

Reply:
We also need to be sure that this logic is only applicable if `CreationPolicy == Create`. Otherwise we don`t want to delete the secrets even if the name changes.

Reply:
Hmm... We could as well, since we know that we will have only one referred `Secret`, to store the name of the generated secret in the `.status` of the `ExternalSecret` and refer to it for the deletion in case that the name of the new generated secret is different.

Reply:
> Hmm... We could as well, since we know that we will have only one referred Secret, to store the name of the generated secret in the .status of the ExternalSecret and refer to it for the deletion in case that the name of the new generated secret is different.
That's a nice idea! :smile: It would be cleaner to find and to delete afterwards (and also to impose the CreationPolicy rules on top of them)